File: backend/app/core/indicators.py
"""Market Intelligence — technical indicators computed in plain code,
OUTSIDE any LLM. The agent feeds these as structured context to the model
(or uses them directly), which is far more reliable than asking an LLM to
do math. Pure functions only — fully unit-testable."""

from __future__ import annotations

from typing import Optional, Sequence


def sma(values: Sequence[float], period: int) -> list[Optional[float]]:
    out: list[Optional[float]] = []
    for i in range(len(values)):
        out.append(None if i + 1 < period else sum(values[i + 1 - period : i + 1]) / period)
    return out


def ema(values: Sequence[float], period: int) -> list[Optional[float]]:
    """EMA seeded with the SMA of the first `period` values (classic convention)."""
    out: list[Optional[float]] = []
    k = 2.0 / (period + 1)
    e: Optional[float] = None
    for i, _v in enumerate(values):
        if i + 1 < period:
            out.append(None)
            continue
        e = sum(values[:period]) / period if e is None else values[i] * k + e * (1 - k)
        out.append(e)
    return out


def rsi(values: Sequence[float], period: int = 14) -> Optional[float]:
    """Wilder's RSI. Returns None when there is not enough data."""
    if len(values) < period + 1:
        return None
    gains = losses = 0.0
    for i in range(1, period + 1):
        d = values[i] - values[i - 1]
        gains += max(d, 0.0)
        losses += max(-d, 0.0)
    ag, al = gains / period, losses / period
    for i in range(period + 1, len(values)):
        d = values[i] - values[i - 1]
        ag = (ag * (period - 1) + max(d, 0.0)) / period
        al = (al * (period - 1) + max(-d, 0.0)) / period
    if al == 0:
        return 100.0
    rs = ag / al
    return 100.0 - 100.0 / (1.0 + rs)


def _last_valid(series: list[Optional[float]]) -> Optional[float]:
    for v in reversed(series):
        if v is not None:
            return v
    return None


def macd(values: Sequence[float], fast: int = 12, slow: int = 26, signal: int = 9) -> dict:
    ef, es = ema(values, fast), ema(values, slow)
    line = [None if (a is None or b is None) else a - b for a, b in zip(ef, es)]
    compact = [v for v in line if v is not None]
    sig = ema(compact, signal) if len(compact) >= signal else [None] * len(compact)
    m = _last_valid(line)
    s = _last_valid(sig)
    hist = None if (m is None or s is None) else m - s
    return {"macd": m, "signal": s, "hist": hist}


def bollinger(values: Sequence[float], period: int = 20, k: float = 2.0) -> dict:
    if len(values) < period:
        return {"mid": None, "upper": None, "lower": None}
    window = values[-period:]
    mid = sum(window) / period
    var = sum((x - mid) ** 2 for x in window) / period
    sd = var ** 0.5
    return {"mid": mid, "upper": mid + k * sd, "lower": mid - k * sd}


def atr(highs: Sequence[float], lows: Sequence[float], closes: Sequence[float], period: int = 14) -> Optional[float]:
    n = len(closes)
    if n < period + 1:
        return None
    trs: list[float] = []
    for i in range(1, n):
        tr = max(
            highs[i] - lows[i],
            abs(highs[i] - closes[i - 1]),
            abs(lows[i] - closes[i - 1]),
        )
        trs.append(tr)
    a = sum(trs[:period]) / period
    for tr in trs[period:]:
        a = (a * (period - 1) + tr) / period
    return a


def snapshot(bars: Sequence[dict]) -> dict:
    """One structured indicator snapshot from OHLCV bars — the exact payload
    the agent gives the LLM (or reasons over itself). Includes ATR-based
    stop/target levels (2x risk / 3x reward)."""
    closes = [b["c"] for b in bars]
    highs = [b["h"] for b in bars]
    lows = [b["l"] for b in bars]
    last = closes[-1]
    prev = closes[-2] if len(closes) > 1 else last
    e20, e50 = _last_valid(ema(closes, 20)), _last_valid(ema(closes, 50))
    r = rsi(closes)
    m = macd(closes)
    bb = bollinger(closes)
    a = atr(highs, lows, closes)
    stop = last - 2 * a if a else last * 0.98
    target = last + 3 * a if a else last * 1.03
    return {
        "last": round(last, 2),
        "change_pct": round((last / prev - 1) * 100, 2) if prev else 0.0,
        "rsi": round(r, 1) if r is not None else None,
        "ema20": round(e20, 2) if e20 else None,
        "ema50": round(e50, 2) if e50 else None,
        "trend": "UP" if (e20 and e50 and e20 > e50) else "DOWN",
        "macd": round(m["macd"], 3) if m["macd"] is not None else None,
        "macd_hist": round(m["hist"], 3) if m["hist"] is not None else None,
        "bb_upper": round(bb["upper"], 2) if bb["upper"] else None,
        "bb_lower": round(bb["lower"], 2) if bb["lower"] else None,
        "atr": round(a, 2) if a else None,
        "stop": round(stop, 2),
        "target": round(target, 2),
    }

File: mobile/lib/core/engine/indicators.dart
/// Market Intelligence — technical indicators computed in plain Dart,
/// OUTSIDE any LLM. Port of `backend/app/core/indicators.py`. The agent
/// feeds these as structured context to the model (or reasons over them
/// directly), which is far more reliable than asking an LLM to do math.
///
/// Pure functions only — fully unit-testable.
library;

import '../models.dart';

List<double?> sma(List<double> values, int period) {
  final out = List<double?>.filled(values.length, null);
  for (var i = period - 1; i < values.length; i++) {
    var sum = 0.0;
    for (var j = i + 1 - period; j <= i; j++) {
      sum += values[j];
    }
    out[i] = sum / period;
  }
  return out;
}

/// EMA seeded with the SMA of the first `period` values (classic convention).
List<double?> ema(List<double> values, int period) {
  final out = List<double?>.filled(values.length, null);
  if (values.length < period) return out;
  final k = 2.0 / (period + 1);
  double? e;
  for (var i = 0; i < values.length; i++) {
    if (i + 1 < period) continue;
    if (e == null) {
      var sum = 0.0;
      for (var j = 0; j < period; j++) {
        sum += values[j];
      }
      e = sum / period;
    } else {
      e = values[i] * k + e * (1 - k);
    }
    out[i] = e;
  }
  return out;
}

/// Wilder's RSI. Returns null when there is not enough data.
double? rsi(List<double> values, [int period = 14]) {
  if (values.length < period + 1) return null;
  var gains = 0.0, losses = 0.0;
  for (var i = 1; i <= period; i++) {
    final d = values[i] - values[i - 1];
    gains += d > 0 ? d : 0.0;
    losses += d < 0 ? -d : 0.0;
  }
  var ag = gains / period, al = losses / period;
  for (var i = period + 1; i < values.length; i++) {
    final d = values[i] - values[i - 1];
    ag = (ag * (period - 1) + (d > 0 ? d : 0.0)) / period;
    al = (al * (period - 1) + (d < 0 ? -d : 0.0)) / period;
  }
  if (al == 0) return 100.0;
  final rs = ag / al;
  return 100.0 - 100.0 / (1.0 + rs);
}

double? _lastValid(List<double?> series) {
  for (var i = series.length - 1; i >= 0; i--) {
    final v = series[i];
    if (v != null) return v;
  }
  return null;
}

class MacdResult {
  const MacdResult({this.macd, this.signal, this.hist});
  final double? macd;
  final double? signal;
  final double? hist;
}

MacdResult macd(
  List<double> values, {
  int fast = 12,
  int slow = 26,
  int signal = 9,
}) {
  final ef = ema(values, fast), es = ema(values, slow);
  final line = List<double?>.generate(values.length, (i) {
    final a = ef[i], b = es[i];
    return (a == null || b == null) ? null : a - b;
  });
  final compact = line.whereType<double>().toList();
  final List<double?> sig;
  if (compact.length >= signal) {
    sig = ema(compact, signal);
  } else {
    sig = List<double?>.filled(compact.length, null);
  }
  double? lastSig;
  for (var i = sig.length - 1; i >= 0; i--) {
    if (sig[i] != null) {
      lastSig = sig[i];
      break;
    }
  }
  final m = _lastValid(line);
  final s = lastSig;
  return MacdResult(
    macd: m,
    signal: s,
    hist: (m == null || s == null) ? null : m - s,
  );
}

class BollingerResult {
  const BollingerResult({this.mid, this.upper, this.lower});
  final double? mid, upper, lower;
}

BollingerResult bollinger(
  List<double> values, [
  int period = 20,
  double k = 2.0,
]) {
  if (values.length < period) {
    return const BollingerResult();
  }
  final window = values.sublist(values.length - period);
  final mid = window.reduce((a, b) => a + b) / period;
  var variance = 0.0;
  for (final x in window) {
    variance += (x - mid) * (x - mid);
  }
  variance /= period;
  final sd = variance <= 0 ? 0.0 : _sqrt(variance);
  return BollingerResult(mid: mid, upper: mid + k * sd, lower: mid - k * sd);
}

double _sqrt(double x) {
  if (x <= 0) return 0;
  var guess = x;
  for (var i = 0; i < 40; i++) {
    guess = 0.5 * (guess + x / guess);
  }
  return guess;
}

/// Average True Range (Wilder smoothing). Null when not enough data.
double? atr(
  List<double> highs,
  List<double> lows,
  List<double> closes, [
  int period = 14,
]) {
  final n = closes.length;
  if (n < period + 1) return null;
  final trs = <double>[];
  for (var i = 1; i < n; i++) {
    trs.add(
      _max3(
        highs[i] - lows[i],
        (highs[i] - closes[i - 1]).abs(),
        (lows[i] - closes[i - 1]).abs(),
      ),
    );
  }
  var a = 0.0;
  for (var i = 0; i < period; i++) {
    a += trs[i];
  }
  a /= period;
  for (var i = period; i < trs.length; i++) {
    a = (a * (period - 1) + trs[i]) / period;
  }
  return a;
}

double _max3(double a, double b, double c) {
  var m = a;
  if (b > m) m = b;
  if (c > m) m = c;
  return m;
}

/// One structured indicator snapshot from OHLCV bars — the exact payload the
/// agent gives the LLM (or reasons over itself). ATR-based stop/target
/// levels (2x risk / 3x reward).
Map<String, dynamic> snapshot(List<Candle> bars) {
  final closes = bars.map((b) => b.close).toList();
  final highs = bars.map((b) => b.high).toList();
  final lows = bars.map((b) => b.low).toList();
  final last = closes.last;
  final prev = closes.length > 1 ? closes[closes.length - 2] : last;
  final e20 = _lastValid(ema(closes, 20));
  final e50 = _lastValid(ema(closes, 50));
  final r = rsi(closes);
  final m = macd(closes);
  final bb = bollinger(closes);
  final a = atr(highs, lows, closes);
  final stop = a != null ? last - 2 * a : last * 0.98;
  final target = a != null ? last + 3 * a : last * 1.03;
  return {
    'last': _round2(last),
    'change_pct': prev > 0 ? _round2((last / prev - 1) * 100) : 0.0,
    'rsi': r == null ? null : _round1(r),
    'ema20': e20 == null ? null : _round2(e20),
    'ema50': e50 == null ? null : _round2(e50),
    'trend': (e20 != null && e50 != null && e20 > e50) ? 'UP' : 'DOWN',
    'macd': m.macd == null ? null : _round3(m.macd!),
    'macd_hist': m.hist == null ? null : _round3(m.hist!),
    'bb_upper': bb.upper == null ? null : _round2(bb.upper!),
    'bb_lower': bb.lower == null ? null : _round2(bb.lower!),
    'atr': a == null ? null : _round2(a),
    'stop': _round2(stop),
    'target': _round2(target),
  };
}

double _round2(double v) => (v * 100).roundToDouble() / 100;
double _round1(double v) => (v * 10).roundToDouble() / 10;
double _round3(double v) => (v * 1000).roundToDouble() / 1000;
